LibraryInstrumentationGuidance.md

```diff
@@ -0,0 +1,115 @@
+
+```csharp
+    static DiagnosticListener source = new DiagnosticListener("Example.MyLibrary");
+
+    Activity activity = null;
+    // create activity only when requested
+    if (source.IsEnabled() && source.IsEnabled("Example.MyLibrary.MyOperation"))
+    {
+        activity = new Activity("Example.MyLibrary.MyOperation");
+        source.StartActivity(activity, new { Input = input });
+    }
+
+    object output = null;
+    try
+    {
+        // perform the actual operation
+        output = RunOperation(input);  
+    }
+    finally
+    {
+        // stop activity if started
+        if (activity != null)
+             source.StopActivity(activity, new { Input = input, Output = output }); 
+    }
+```
+> ### *__TODO__ - provide a pointer to a document with more advanced instrumentation examples (WIP)* 
+
+### Payload
+
+When starting and stopping an activity, it is possible to pass a payload object, which will be available for the listener. This way the listener can access raw, non-serialized data in real time. It can be useful for pulling out additional diagnostic information but also manipulating data as it is being processed (for example, inject diagnostic context before an outbound call is made). Good examples of such payload are ```HttpRequestMessage``` or messages that are being passed through queues.
+
+Mind that payload is not preserved as part of Activity and is only available when activity is started/stopped. Therefore it is a good practice to specify all data that was passed to ```StartActivity()``` in ```StopActivity()``` as well.
+
+#### Payload format
+
+Diagnostic source event and activity start/stop API allows to specify only a single payload object. However, in order to pass more data and allow future additions the recommendation is to use dynamic objects. Since these are .NET objects the names of particular properties should follow [standard .NET naming convention][DotNetPropertyNamingConvention]. 
+
+Here are some recommendations for typical payload property names:
+
+| Property name | Description |
+|:--------------|:-------------------|
+| `Endpoint` | The ```Uri``` of an endpoint the activity is for (for example, target database, service) |
+| `PartitionKey` | The key/ID of the partition the activity is for |
+| `Status` | The ```TaskStatus``` of a completed asynchronous task |
+| `Exception` | The captured exception object |
+
+### Performance
+
+In order to avoid unnecessary overhead, it is highly recommended to check if there is any listener for given activity: 
+
+```csharp
+    source.IsEnabled() && source.IsEnabled("Example.MyLibrary.MyOperation")
+```
+
+The parameterless ```source.IsEnabled()``` check should be put before any other as it is very efficient and can virtually prevent any overhead in absence of any listener for given diagnostic source.
```

Comment:
Is it too late to modify DiagnosticSource to incorporate the functionality of the parameterless IsEnabled() within the overload that takes a string? i.e. the first line of `bool IsEnabled(string activityName)` does a short-circuit test: `if (!IsEnabled()) return false;`

Reply:
Oh, I see. You are trying to avoid the virtual call.